File: app/src/main/java/ch/unstable/ost/error/model/ErrorInfo.java
package ch.unstable.ost.error.model;


import android.support.annotation.Nullable;

import java.io.PrintWriter;
import java.io.StringWriter;
import java.util.Date;

public class ErrorInfo {

    public static final ErrorInfo EMPTY = new ErrorInfo((String) null);

    @Nullable
    private final Throwable exception;
    @Nullable
    private final String stackTrace;

    private final Date timestamp = new Date();

    public ErrorInfo(@Nullable String stackTrace) {
        this.exception = null;
        this.stackTrace = stackTrace;
    }

    public ErrorInfo(@Nullable Throwable exception) {
        this.exception = exception;
        this.stackTrace = null;
    }

    @Nullable
    public Throwable getException() {
        return exception;
    }

    public String getStackTrace() {
        if (exception == null && stackTrace == null) {
            return "";
        } else if(stackTrace != null) {
            return stackTrace;
        } else {
            StringWriter sw = new StringWriter();
            PrintWriter pw = new PrintWriter(sw);
            exception.printStackTrace(pw);
            return sw.toString();
        }
    }

    public Date getTimestamp() {
        return timestamp;
    }
}

File: app/src/main/java/ch/unstable/ost/error/model/ErrorInfo.kt
package ch.unstable.ost.error.model

import java.io.PrintWriter
import java.io.StringWriter
import java.util.*

class ErrorInfo(private val stackTrace: String?) {
    val timestamp = Date()
    val exception: Throwable? = null

    fun getStackTrace(): String {
        return if (exception == null && stackTrace == null) {
            ""
        } else if (stackTrace != null) {
            stackTrace
        } else {
            val sw = StringWriter()
            val pw = PrintWriter(sw)
            exception!!.printStackTrace(pw)
            sw.toString()
        }
    }

    companion object {
        @JvmField
        val EMPTY = ErrorInfo(null as String?)
    }
}